### PR TITLE
rust/kernel: move from_kernel_result! macro to error.rs

### PR DESCRIPTION
**Pull request by TheSven73**
This macro is well-suited for use elsewhere in the Rust core.
Move it out to a suitable place, error.rs.

(I'd like to use this in #254)

## v1 -> v2
- split up in logically consistent commits (ojeda)
- add commit to improve safety and eliminate possible panics (TheSven73)

### CODE REVIEW

**Review by alex — APPROVED** [review]

File: rust/kernel/error.rs [hunk 1]
```diff
@@ -104,3 +105,44 @@ impl From<AllocError> for Error {
...
 }
+
+pub fn from_kernel_result_helper<T>(r: Result<T>) -> T
+where
+    T: TryFrom<c_types::c_int>,
+    T::Error: core::fmt::Debug,
+{
+    match r {
+        Ok(v) => v,
+        Err(e) => T::try_from(e.to_kernel_errno()).unwrap(),
```
Comment (TheSven73) [inline rust/kernel/error.rs]: Just out of curiosity: could this `unwrap` panic?

Comment (alex) [reply rust/kernel/error.rs]: I don't think so, no.

File: rust/kernel/error.rs [hunk 2]
```diff
@@ -104,3 +105,44 @@ impl From<AllocError> for Error {
         Error::ENOMEM
     }
 }
+
+pub fn from_kernel_result_helper<T>(r: Result<T>) -> T
```
Comment (ojeda) [inline rust/kernel/error.rs:113]: Does this need `doc(hidden)`?

File: rust/kernel/error.rs [hunk 3]
```diff
@@ -104,3 +105,44 @@ impl From<AllocError> for Error {
...
+    T: TryFrom<c_types::c_int>,
+    T::Error: core::fmt::Debug,
+{
+    match r {
+        Ok(v) => v,
+        Err(e) => T::try_from(e.to_kernel_errno()).unwrap(),
+    }
+}
+
+/// Transforms a `crate::error::Result<T>` to a kernel `C` integer result.
```
Comment (ojeda) [inline rust/kernel/error.rs]: C is not code, i.e.:

```suggestion
/// Transforms a `crate::error::Result<T>` to a kernel C integer result.
```

Does an intra-doc link work here?

```suggestion
/// Transforms a [`crate::error::Result<T>`] to a kernel C integer result.
```

File: rust/kernel/error.rs [hunk 4]
```diff
@@ -104,3 +105,44 @@ impl From<AllocError> for Error {
...
+///
+/// This is useful when calling Rust functions that return `Result<T>`
```
Comment (ojeda) [inline rust/kernel/error.rs]: Intra-doc link perhaps?

File: rust/kernel/error.rs [hunk 5]
```diff
@@ -104,3 +105,44 @@ impl From<AllocError> for Error {
...
+/// unsafe extern "C" fn probe_callback(
+///     pdev: *mut bindings::platform_device,
+/// ) -> c_types::c_int {
+///     from_kernel_result! {
+///         let ptr = devm_alloc(pdev)?;
+///         rust_helper_platform_set_drvdata(pdev, ptr);
+///         Ok(0)
+///     }
+/// }
+/// ```
```
Comment (ojeda) [inline rust/kernel/error.rs:146]: In general, please split commits if you are doing several changes. e.g. the documentation addition (which is great!) should be on its own commit.

File: rust/kernel/error.rs (hunk 1 above)
Comment (ojeda) [reply rust/kernel/error.rs]: Good point, please use `NOPANIC` there then.

Comment (TheSven73) [reply rust/kernel/error.rs]: Thinking about this a bit more. We can't put `NOPANIC` here, because I suspect that it **can** panic! Consider an `extern C` function that returns `c_char` for example. A contrived example, but hey.

```rust
unsafe extern "C" fn my_callback() -> c_types::c_char {
    from_kernel_result! {
        some_function()?;
        Ok(0)
    }
}
```
Now imagine that `some_function` returns `Err(-EPROBE_DEFER)` which corresponds to `-517`. Since `T` is now a `c_char`, and `-517` can't fit, the function will panic and the kernel dies a horrible death.

Suggestion: we know that `errno`s will never exceed `MAX_ERRNO=4095`. So our negative error codes will allays fit into an `i16` (we can perhaps add a static assertion to make sure this holds always). And then we can use a `From` trait:

```rust
pub fn from_kernel_result_helper<T>(r: Result<T>) -> T
where
    T: From<i16>,
{
    match r {
        Ok(v) => v,
        Err(e) => T::from(e.to_kernel_errno() as i16),
    }
}
```
Now things will work when returning an integer type that fits (`c_int`, `c_ssize_t`, etc) but `c_char` will fail, which is what we want !
```
error[E0277]: the trait bound `i8: From<i16>` is not satisfied
   --> rust/kernel/error.rs:144:50
    |
110 |   pub fn from_kernel_result_helper<T>(r: Result<T>) -> T
    |          ------------------------- required by a bound in this
111 |   where
112 |       T: From<i16>,
    |          --------- required by this bound in `from_kernel_result_helper`
...
144 |           $crate::error::from_kernel_result_helper((|| {
    |  __________________________________________________^
145 | |             $($tt)*
146 | |         })())
    | |____________^ the trait `From<i16>` is not implemented for `i8`
```

Comment (alex) [reply rust/kernel/error.rs]: SGTM, maybe add a `static_assert!` that `MAX_ERRNO` fits in an `i16`

File: rust/kernel/error.rs [hunk 6]
```diff
@@ -106,15 +106,20 @@ impl From<AllocError> for Error {
     }
 }
 
+// # Invariant: `-bindings::MAX_ERRNO` fits in an `i16`.
+crate::static_assert!(bindings::MAX_ERRNO <= -(i16::MIN as i32) as u32);
```
Comment (TheSven73) [inline rust/kernel/error.rs:110]: I have the nagging feeling that this method is too contrived. Is there an idiomatic way to verify whether a value fits in a type? I tried `Try` or `TryFrom` but that doesn't work, because those types can't run inside a `static_assert!`.

Comment (ojeda) [reply rust/kernel/error.rs:110]: I have taken a look at this -- it is indeed a bit painful. After having tried several approaches, I am preparing a PR with a fun one.

Comment (TheSven73) [reply rust/kernel/error.rs:110]: I am grabbing the popcorn !! :)

Comment (ojeda) [reply rust/kernel/error.rs:110]: See if you like #269 :)

It would allow you to do:

    static_assert!((-(bindings::MAX_ERRNO as i128)) fits in i16);

Comment (TheSven73) [reply rust/kernel/error.rs:110]: I like it a lot !! Will definitely rebase to that after it gets merged !! :100:

**Comment by TheSven73** [comment]
Corrected minor nit in commit message to restart CI (which failed due to inability to download packages).

**Comment by ksquirrel** [comment]
Review of `eb4c6e8f9a75`:
  - ✔️ Commit [`7dbb95f`](https://github.com/Rust-for-Linux/linux/commit/7dbb95fbda6893c17bcf355f5b8e8a76896d93c6): Looks fine!
  - ✔️ Commit [`2ce3372`](https://github.com/Rust-for-Linux/linux/commit/2ce3372a23cd976596504b55f161837a598749e2): Looks fine!
  - ✔️ Commit [`eb4c6e8`](https://github.com/Rust-for-Linux/linux/commit/eb4c6e8f9a75e2a8061d5b47c6a72e6646c3e669): Looks fine!

**Comment by alex** [comment]
Going to merge this now, and we can improve the static_assert in a follow-up PR.
